beta/src/components/MDX/Sandpack/LoadingOverlay.tsx

```diff
@@ -0,0 +1,143 @@
+import {useState} from 'react';
+
+import {
+  LoadingOverlayState,
+  OpenInCodeSandboxButton,
+  useSandpack,
+} from '@codesandbox/sandpack-react';
+import {useEffect} from 'react';
+
+const FADE_ANIMATION_DURATION = 200;
+
+export const LoadingOverlay = ({
+  clientId,
+  dependenciesLoading,
+  forceLoading,
+}: {
+  clientId: string;
+  dependenciesLoading: boolean;
+  forceLoading: boolean;
+} & React.HTMLAttributes<HTMLDivElement>): JSX.Element | null => {
+  const loadingOverlayState = useLoadingOverlayState(
+    clientId,
+    dependenciesLoading,
+    forceLoading
+  );
+
+  if (loadingOverlayState === 'HIDDEN') {
+    return null;
+  }
+
+  if (loadingOverlayState === 'TIMEOUT') {
+    return (
+      <div className="sp-overlay sp-error">
+        <div className="sp-error-message">
+          Unable to establish connection with the sandpack bundler. Make sure
+          you are online or try again later. If the problem persists, please
+          report it via{' '}
+          <a
+            className="sp-error-message"
+            href="mailto:hello@codesandbox.io?subject=Sandpack Timeout Error">
+            email
+          </a>{' '}
+          or submit an issue on{' '}
+          <a
+            className="sp-error-message"
+            href="https://github.com/codesandbox/sandpack/issues"
+            rel="noreferrer noopener"
+            target="_blank">
+            GitHub.
+          </a>
+        </div>
+      </div>
+    );
+  }
+
+  const stillLoading =
+    loadingOverlayState === 'LOADING' || loadingOverlayState === 'PRE_FADING';
+
+  return (
+    <div
+      className="sp-overlay sp-loading"
+      style={{
+        opacity: stillLoading ? 1 : 0,
+        transition: `opacity ${FADE_ANIMATION_DURATION}ms ease-out`,
+      }}>
+      <div className="sp-cube-wrapper" title="Open in CodeSandbox">
+        <OpenInCodeSandboxButton />
+        <div className="sp-cube">
+          <div className="sp-sides">
+            <div className="top" />
+            <div className="right" />
+            <div className="bottom" />
+            <div className="left" />
+            <div className="front" />
+            <div className="back" />
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+const useLoadingOverlayState = (
+  clientId: string,
+  dependenciesLoading: boolean,
+  forceLoading: boolean
+): LoadingOverlayState => {
+  const {sandpack, listen} = useSandpack();
+  const [state, setState] = useState<LoadingOverlayState>('LOADING');
+
+  if (state !== 'LOADING' && forceLoading) {
+    setState('LOADING');
+  }
+
+  /**
+   * Sandpack listener
+   */
+  useEffect(() => {
+    sandpack.loadingScreenRegisteredRef.current = true;
+
+    const unsubscribe = listen((message) => {
+      if (message.type === 'done') {
+        setState((prev) => {
+          return prev === 'LOADING' ? 'PRE_FADING' : 'HIDDEN';
+        });
+      }
+    }, clientId);
+
+    return (): void => {
+      unsubscribe();
```

Comment:
should this unset the ref you set in setup function?

Reply:
Humm I'm not sure I got what you mean. This should only unsubscribe this listener and nothing else, so all other listeners and initial iframe ref should stay untouched.

Reply:
i mean that the cleanup function should usually be fully symmetric to the setup function, so i’m curious what the ref you’re setting on the first line represents, and whether that should be “undone” too.

the idea is that we’ll add a mode in the which effects get unmounted but state/refs get preserved. like unmounting that preserves state/DOM (just hiding the nodes visually). in that world, would it make sense to reset the ref to false? that’s what i mean by it being symmetric.

Reply:
Ah got it! I removed this ref on the previous commit, as this is a piece from the legacy CodeSandbox bundler and doesn't do anything on the new bundler. 

Thanks for the explanation. So, answering your question and considering the scenario you described, it'd make sense to preserve the ref value on unmounted. Because this ref passes a variable to the bundler that stays mounted regardless of the usage of this hook.